locations/middlewares/anti_bot_detection.py

```diff
@@ -0,0 +1,103 @@
+from enum import Enum
+
+from scrapy import Request, Spider
+from scrapy.downloadermiddlewares.httpcompression import HttpCompressionMiddleware
+from scrapy.http import HtmlResponse, Response
+
+
+# As a temporary transition to use of AntiBotDetectionMiddleware,
+# assume that Zyte API is capable of bypassing all anti-bot
+# methods. This list will need to be amended after testing of Zyte
+# API ability to bypass various anti-bot methods. This list will
+# need to be frequently updated depending on the effectiveness of
+# Zyte API's ability to rapidly bypass new/revised anti-bot methods.
+class AntiBotMethods(Enum):
+    # Azure Web Applicatigon Firewall bot protection documentation:
+    # https://learn.microsoft.com/en-us/azure/web-application-firewall/ag/bot-protection-overview
+    AZURE_WAF = {"name": "Azure WAF", "zyte_bypassable": True}
+
+    # Cloudflare bot protection documentation:
+    # https://developers.cloudflare.com/bots/
+    CLOUDFLARE = {"name": "Cloudflare", "zyte_bypassable": True}
+
+    # DataDome bot protection product brochure:
+    # https://datadome.co/products/bot-protection/
+    DATADOME = {"name": "DataDome", "zyte_bypassable": True}
+
+    # HUMAN Bot Defender bot protection product brochure:
+    # https://www.humansecurity.com/products/human-bot-defender
+    HUMAN = {"name": "HUMAN", "zyte_bypassable": True}
+
+    # Imperva bot protection product brochure:
+    # https://www.imperva.com/products/bot-detection-mitigation/
+    IMPERVA = {"name": "Imperva", "zyte_bypassable": True}
+
+    # Qrator bot protection documentation:
+    # https://docs.qrator.net/en/technologies/tracking-cookie.html
+    QRATOR = {"name": "Qrator", "zyte_bypassable": True}
+
+
+class AntiBotDetectionMiddleware:
+    def __init__(self, crawler):
+        self.crawler = crawler
+
+    @classmethod
+    def from_crawler(cls, crawler):
+        return cls(crawler)
+
+    def process_response(self, request: Request, response: Response, spider: Spider):
+        if response.headers.get("cf-mitigated"):
+            self.add_anti_bot_method(AntiBotMethods.CLOUDFLARE, spider)
+
+        if response.headers.get("x-datadome"):
+            self.add_anti_bot_method(AntiBotMethods.DATADOME, spider)
+
+        if server := response.headers.get("server"):
+            server_utf8 = self.decode_http_header_value(server)
+            if server_utf8.upper() == "QRATOR":
+                self.add_anti_bot_method(AntiBotMethods.QRATOR, spider)
+            if response.status == 403 and server_utf8.upper() == "MICROSOFT-AZURE-APPLICATION-GATEWAY/V2":
+                self.add_anti_bot_method(AntiBotMethods.AZURE_WAF, spider)
+
+        if cookies := response.headers.getlist("set-cookie"):
```

Comment:
I think scrapy is having a discussion about making cookies more accessible, I'll try and find the link tomorrow. We may want to be part of that conversation

Reply:
Is it https://github.com/scrapy/scrapy/issues/5431 ?

Reply:
Or https://github.com/scrapy/scrapy/pull/6218